File: Dynamic-Programming/152. Maximum Product Subarray/maximumProductSubarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        cur_min = nums[0]
        cur_max = nums[0]
        final_max = nums[0]
        
        for i in range(1, len(nums)):
            # when nums[i] < 0
            # the future cur_max is cur_min
            # vice versa
            if nums[i] < 0:
                temp = cur_max
                cur_max = cur_min
                cur_min = temp

            cur_max = max(cur_max * nums[i], nums[i])
            cur_min = min(cur_min * nums[i], nums[i])

            final_max = max(cur_max, final_max)
        
        return final_max

File: Dynamic-Programming/152. Maximum Product Subarray/maximumProductSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxProduct = function(nums) {
    let curMax = nums[0];
    let curMin = nums[0];
    let max = nums[0];
    
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] < 0) {
            let temp = curMax;
            curMax = curMin;
            curMin = temp;
        }
        
        curMax = Math.max(curMax * nums[i], nums[i]);
        curMin = Math.min(curMin * nums[i], nums[i]);
        max = Math.max(curMax, max);
    }
    
    return max;
};